python/hopsworks_common/core/dataset_api.py

```diff
@@ -158,44 +159,93 @@ def upload(
 
         dataset_api = project.get_dataset_api()
 
+        # upload a file to Resources dataset
         uploaded_file_path = dataset_api.upload("my_local_file.txt", "Resources")
 
+        # upload a directory to Resources dataset
+        uploaded_file_path = dataset_api.upload("my_dir", "Resources")
+
         ```
         # Arguments
-            local_path: local path to file to upload
+            local_path: local path to file or directory to upload, can be relative or absolute
             upload_path: path to directory where to upload the file in Hopsworks Filesystem
-            overwrite: overwrite file if exists
+            overwrite: overwrite file or directory if exists
             chunk_size: upload chunk size in bytes. Default 1048576 bytes
-            simultaneous_uploads: number of simultaneous chunks to upload. Default 3
+            simultaneous_chunks: number of simultaneous chunks to upload for each file upload. Default 3
+            simultaneous_uploads: number of simultaneous files to be uploaded for directories. Default 3
             max_chunk_retries: maximum retry for a chunk. Default is 1
             chunk_retry_interval: chunk retry interval in seconds. Default is 1sec
         # Returns
-            `str`: Path to uploaded file
+            `str`: Path to uploaded file or directory
         # Raises
-            `RestAPIError`: If unable to upload the file
+            `RestAPIError`: If unable to upload the file or directory
         """
+
         # local path could be absolute or relative,
         if not os.path.isabs(local_path) and os.path.exists(
             os.path.join(os.getcwd(), local_path)
         ):
             local_path = os.path.join(os.getcwd(), local_path)
 
-        file_size = os.path.getsize(local_path)
-
         _, file_name = os.path.split(local_path)
 
         destination_path = upload_path + "/" + file_name
 
         if self.exists(destination_path):
             if overwrite:
-                self.remove(destination_path)
+                if 'datasetType' in self._get(destination_path):
+                    raise DatasetException("overwrite=True not supported on a top-level dataset")
+                else:
+                    self.remove(destination_path)
             else:
                 raise DatasetException(
                     "{} already exists, set overwrite=True to overwrite it".format(
-                        local_path
+                        destination_path
                     )
                 )
 
+        if os.path.isdir(local_path):
+            self.mkdir(destination_path)
+
+        if os.path.isdir(local_path):
+            with ThreadPoolExecutor(simultaneous_uploads) as executor:
```

Comment:
Just being a bit picky, I would also extract the code for uploading a directory (`ThreadPoolExecutor`) to a separate `_upload_dir_files` method